File: kotlin-desktop-toolkit/src/main/kotlin/org/jetbrains/desktop/macos/Converters.kt
package org.jetbrains.desktop.macos

import org.jetbrains.desktop.macos.generated.NativeAutoDropArray_AutoDropArray_u8
import org.jetbrains.desktop.macos.generated.NativeAutoDropArray_RustAllocatedStrPtr
import org.jetbrains.desktop.macos.generated.NativeAutoDropArray_u8
import org.jetbrains.desktop.macos.generated.NativeBorrowedArray_BorrowedStrPtr
import org.jetbrains.desktop.macos.generated.NativeBorrowedArray_u8
import org.jetbrains.desktop.macos.generated.NativeColor
import org.jetbrains.desktop.macos.generated.NativeLogicalPoint
import org.jetbrains.desktop.macos.generated.NativeLogicalRect
import org.jetbrains.desktop.macos.generated.NativeLogicalSize
import org.jetbrains.desktop.macos.generated.NativePhysicalSize
import org.jetbrains.desktop.macos.generated.desktop_macos_h.NativeBorrowedStrPtr
import java.lang.foreign.Arena
import java.lang.foreign.MemorySegment
import java.lang.foreign.ValueLayout.JAVA_BYTE

internal fun LogicalSize.Companion.fromNative(s: MemorySegment): LogicalSize {
    return LogicalSize(
        width = NativeLogicalSize.width(s),
        height = NativeLogicalSize.height(s),
    )
}

internal fun LogicalSize.toNative(arena: Arena): MemorySegment {
    val result = NativeLogicalSize.allocate(arena)
    NativeLogicalSize.width(result, width)
    NativeLogicalSize.height(result, height)
    return result
}

internal fun LogicalPoint.Companion.fromNative(s: MemorySegment): LogicalPoint {
    return LogicalPoint(
        x = NativeLogicalPoint.x(s),
        y = NativeLogicalPoint.y(s),
    )
}

internal fun LogicalPoint.toNative(arena: Arena): MemorySegment {
    val result = NativeLogicalPoint.allocate(arena)
    NativeLogicalPoint.x(result, x)
    NativeLogicalPoint.y(result, y)
    return result
}

internal fun PhysicalSize.Companion.fromNative(s: MemorySegment): PhysicalSize {
    return PhysicalSize(
        width = NativePhysicalSize.width(s),
        height = NativePhysicalSize.height(s),
    )
}

internal fun PhysicalSize.toNative(arena: Arena): MemorySegment {
    val result = NativePhysicalSize.allocate(arena)
    NativePhysicalSize.width(result, width)
    NativePhysicalSize.height(result, height)
    return result
}

internal fun LogicalRect.toNative(s: MemorySegment) {
    NativeLogicalPoint.x(NativeLogicalRect.origin(s), origin.x)
    NativeLogicalPoint.y(NativeLogicalRect.origin(s), origin.y)
    NativeLogicalSize.width(NativeLogicalRect.size(s), size.width)
    NativeLogicalSize.height(NativeLogicalRect.size(s), size.height)
}

// internal fun PhysicalPoint.Companion.fromNative(s: MemorySegment): PhysicalPoint {
//    return PhysicalPoint(x = NativePhysicalPoint.x(s),
//                 y = NativePhysicalPoint.y(s))
// }

internal fun Color.toNative(arena: Arena): MemorySegment {
    val result = NativeColor.allocate(arena)
    NativeColor.red(result, red)
    NativeColor.green(result, green)
    NativeColor.blue(result, blue)
    NativeColor.alpha(result, alpha)
    return result
}

internal fun listOfStringsFromNative(segment: MemorySegment): List<String> {
    val ptr = NativeAutoDropArray_RustAllocatedStrPtr.ptr(segment)
    val len = NativeAutoDropArray_RustAllocatedStrPtr.len(segment)

    return (0 until len).map { i ->
        val strPtr = ptr.getAtIndex(NativeAutoDropArray_RustAllocatedStrPtr.`ptr$layout`(), i)
        strPtr.getUtf8String(0)
    }.toList()
}

internal fun listOfStringsToNative(arena: Arena, list: List<String>): MemorySegment {
    val itemsCount = list.count().toLong()
    val itemsArray = arena.allocateArray(NativeBorrowedStrPtr, itemsCount)
    list.forEachIndexed { index, item ->
        val strPtr = arena.allocateUtf8String(item)
        itemsArray.setAtIndex(NativeBorrowedStrPtr, index.toLong(), strPtr)
    }
    val result = NativeBorrowedArray_BorrowedStrPtr.allocate(arena)
    NativeBorrowedArray_BorrowedStrPtr.ptr(result, itemsArray)
    NativeBorrowedArray_BorrowedStrPtr.len(result, itemsCount)
    return result
}

internal fun ByteArray.toNative(arena: Arena): MemorySegment = let { bytes ->
    val result = NativeBorrowedArray_u8.allocate(arena)
    NativeBorrowedArray_u8.ptr(result, arena.allocateArray(JAVA_BYTE, *bytes))
    NativeBorrowedArray_u8.len(result, bytes.count().toLong())
    result
}

internal fun byteArrayFromNative(segment: MemorySegment): ByteArray {
    val ptr = NativeAutoDropArray_u8.ptr(segment)
    val len = NativeAutoDropArray_u8.len(segment)
    val result = ByteArray(len.toInt())
    MemorySegment.copy(ptr, JAVA_BYTE, 0, result, 0, result.size)
    return result
}

internal fun listOfByteArraysFromNative(segment: MemorySegment): List<ByteArray> {
    val ptr = NativeAutoDropArray_AutoDropArray_u8.ptr(segment)
    val len = NativeAutoDropArray_AutoDropArray_u8.len(segment)
    return (0 until len).map { i ->
        val byteArrayPtr = NativeAutoDropArray_u8.asSlice(ptr, i)
        byteArrayFromNative(byteArrayPtr)
    }.toList()
}

File: native/desktop-macos/src/macos/converters.rs
use objc2::rc::Retained;
use objc2_app_kit::NSColor;
use objc2_foundation::{NSPoint, NSRect, NSSize};

use crate::geometry::{Color, LogicalPixels, LogicalPoint, LogicalRect, LogicalSize, PhysicalPoint, PhysicalSize};

impl From<NSSize> for LogicalSize {
    fn from(value: NSSize) -> Self {
        Self {
            width: value.width,
            height: value.height,
        }
    }
}

impl From<NSSize> for PhysicalSize {
    fn from(value: NSSize) -> Self {
        Self {
            width: value.width,
            height: value.height,
        }
    }
}

impl From<NSPoint> for PhysicalPoint {
    fn from(value: NSPoint) -> Self {
        Self { x: value.x, y: value.y }
    }
}

impl From<PhysicalPoint> for NSPoint {
    fn from(value: PhysicalPoint) -> Self {
        NSPoint { x: value.x, y: value.y }
    }
}

impl LogicalPoint {
    #[allow(dead_code)]
    pub(crate) fn as_macos_coords(&self, height: LogicalPixels) -> NSPoint {
        NSPoint {
            x: self.x,
            y: height - self.y,
        }
    }

    pub(crate) fn from_macos_coords(value: NSPoint, height: LogicalPixels) -> Self {
        Self {
            x: value.x,
            y: height - value.y,
        }
    }
}

impl From<LogicalSize> for NSSize {
    fn from(value: LogicalSize) -> Self {
        NSSize {
            width: value.width,
            height: value.height,
        }
    }
}

impl From<PhysicalSize> for NSSize {
    fn from(value: PhysicalSize) -> Self {
        NSSize {
            width: value.width,
            height: value.height,
        }
    }
}

impl LogicalRect {
    pub(crate) fn as_macos_coords(&self, height: LogicalPixels) -> NSRect {
        let origin = NSPoint {
            x: self.origin.x,
            y: height - (self.origin.y + self.size.height),
        };
        NSRect {
            origin,
            size: self.size.into(),
        }
    }

    pub(crate) fn from_macos_coords(value: NSRect, height: LogicalPixels) -> Self {
        let origin = LogicalPoint {
            x: value.origin.x,
            y: height - (value.origin.y + value.size.height),
        };
        Self {
            origin,
            size: value.size.into(),
        }
    }
}

impl From<Color> for Retained<NSColor> {
    fn from(value: Color) -> Self {
        unsafe { NSColor::colorWithSRGBRed_green_blue_alpha(value.red, value.green, value.blue, value.alpha) }
    }
}
